File: LectureCodes/NumQuad/fourcoeffcomp/Matlab/fourcoeffcomp.m
function y = fourcoeffcomp(c,m,ovsmpl)
% Compute the Fourier coefficients \Blue{$y_{-m},\ldots,y_{m}$} of the function
% \Blue{$c:[0,1[\mapsto \bbC$} using an \Magenta{oversampling factor} \texttt{ovsmpl}.
% \texttt{c} must be a handle to a function \texttt{\symbol{64}(t)}, which accepts row
% vector arguments
if (nargin < 3), ovsmpl = 2; else ovsmpl = ceil(ovsmpl); end
N = (2*m+1)*ovsmpl; h = 1/N; % Number of quadrature points
% (Inverse) discrete Fourier transform
y = ifft(c(0:h:1-h));
% Undo oversampling and wrapping of Fourier coefficient array
y = [y(N-m+1:N),y(1:m+1)];

File: LectureCodes/NumQuad/fourcoeffcomp/Python/fourcoeffcomp.py
import numpy as np


def fourcoeffcomp(c, m, ovsmpl=2):
    """ Compute the Fourier coefficients of the function c """
    ovsmpl = np.ceil(ovsmpl)
    # Number of quadrature points
    n = (2 * m + 1) * ovsmpl
    # Inverse DFT
    y = np.fft.ifft(c(np.linspace(0, 1, n, endpoint=False)))
    # Undo oversampling and wrapping of Fourier coefficient array
    return np.hstack([y[n - m:n], y[:m + 1]])


def main():
    y = fourcoeffcomp(lambda x: np.exp(1j * x * np.pi), 4)
    print(y)


if __name__ == '__main__':
    main()
